src/wallet/wallet.cpp

```diff
@@ -260,6 +261,69 @@ std::shared_ptr<CWallet> LoadWalletInternal(WalletContext& context, const std::s
         return nullptr;
     }
 }
+
+class FastWalletRescanFilter
+{
+public:
+    FastWalletRescanFilter(const CWallet& wallet) : m_wallet(wallet)
```

Comment:
Sounds like an interesting idea that is worth to be followed-up, primarily for the reusability argument. I think in the end we'd not get rid of the `assert` and `dynamic_cast` though, instead it's just shifted to the caller? (I don't see a way of avoiding the cast, since the wallet class methods `GetAllScriptPubKeyMans()` and `GetActiveScriptPubKeyMans()` only give us the SPKmans with the base class interface `ScriptPubKeyMan*`).

Reply:
While I writing the suggestion, I implicitly assumed that a CWallet rescan function would call rescan on individual DescriptorScriptPubKeyMans, which would have remove the need for any casting. But scanning is done globally, which makes sense because otherwise you'd have to scan the blocks many times over.

A larger refactor might add `MaybeMatch()` to the SPKM interface, which would return `true` for legacy and would use the filter for descriptors. But this would mean having N separate filters. Not sure how that impacts the overall false-positive rate.

Reply:
> A larger refactor might add MaybeMatch() to the SPKM interface, which would return true for legacy and would use the filter for descriptors. But this would mean having N separate filters. Not sure how that impacts the overall false-positive rate.

I actually have this coded in a separated branch for an on-going project that have (will push the milestone PR soon..).
But essentially, Each spkm have their own elements set, and the update is signal-based instead of have to poll at every block.

I think that, as we will get another use case for it, it might be better to add it in follow-up PR.